### PR TITLE
Add a tool to recreate a deleted cluster from existing PersistentVolumeClaims

### PR DESCRIPTION
**Pull request by sebgl**
Add a tool called `reattach-pv` into our `hack` directory.

It allows recreating a deleted Elasticsearch cluster while reusing existing PersistentVolumes. This is intended for catastrophic failures (or accidental cluster removal) that are not supposed to happen under normal conditions. More details in `README.md`.

Also add a basic test to check the script behaves as expected when targeting a GKE cluster.

And include compilation of this tool as part of CI: since it relies on "internal" operator code, we may accidentally break it if we change the operator code.

### CODE REVIEW

File: hack/reattach-pv/README.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+Recreate an Elasticsearch cluster by reattaching existing released PersistentVolumes
+
+Usage:
+  reattach-pv [flags]
+
+Flags:
+      --dry-run                         do not apply any Kubernetes resource change
+      --elasticsearch-manifest string   path pointing to the Elasticsearch yaml manifest
+  -h, --help                            help for reattach-pv
+      --pv-backup-path string           path to the file where a backup of existing PersistentVolumes will be stored before update, set empty to disable (default "pv_backup_{timestamp}.json")
```
Comment (charith-elastic) [inline hack/reattach-pv/README.md]: To me, the wording here implies that the actual contents of the volumes are copied to make a backup of the data. However, the tool only dumps the PV definition in JSON to a file. Is there any benefit to having the definition backed up?  

In my mind, the risk here is that the restoration operation ends up corrupting the data already stored in the volumes. The only way to mitigate that is to clone the disks (gcloud supports snapshots, not sure about other providers) before the tool is run.

File: hack/reattach-pv/README.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+```
+
+Example:
+
+```
+# perform a dry run first
+./reattach-pv --elasticsearch-manifest elasticsearch.yml --dry-run
+# inspect stdout, and notice a backup of existing volumes has been created in the current directory
```
[the remark this answers is not in the document]
Comment (sebgl) [reply hack/reattach-pv/README.md]: the backup is also created in dry-run mode

File: hack/reattach-pv/main.go [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+
+// esFromFile parses an Elasticsearch resource from the given yaml manifest path.
+func esFromFile(path string) (esv1.Elasticsearch, error) {
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		return esv1.Elasticsearch{}, err
+	}
+	obj, _, err := scheme.Codecs.UniversalDeserializer().Decode(data, nil, nil)
+	if err != nil {
+		return esv1.Elasticsearch{}, nil
```
[the remark this answers is not in the document]
Comment (sebgl) [reply hack/reattach-pv/main.go]: indeed!

File: hack/reattach-pv/main.go [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+	}
+	es := *obj.(*esv1.Elasticsearch)
+	if es.Namespace == "" {
+		fmt.Println("Setting Elasticsearch namespace to 'default'")
+		es.Namespace = "default"
+	}
```
[the remark this answers is not in the document]
Comment (sebgl) [reply hack/reattach-pv/main.go]: That's a good idea. I quickly checked and it does not seem straightforward to retrieve the default namespace from the config attached to our k8s client.
Not sure if it's worth the trouble? The tool will just error out if expected PersistentVolumes don't exist in the default namespace, in which case the user can just specify the namespace in the ES resource.

File: hack/reattach-pv/main.go [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+	return k8s.WrapClient(c), nil
+}
+
+// checkElasticsearchNotFound returns an error if the given Elasticsearch resource already exists.
+func checkElasticsearchNotFound(c k8s.Client, es esv1.Elasticsearch) error {
+	var retrieved esv1.Elasticsearch
+	err := c.Get(k8s.ExtractNamespacedName(&es), &retrieved)
+	if err == nil {
+		return fmt.Errorf("elasticsearch resource %s exists in the apiserver: it should be deleted first", es.Name)
```
[the remark this answers is not in the document]
Comment (sebgl) [reply hack/reattach-pv/main.go]: I changed the second part to `this tool can only recover clusters that don't exist anymore`

File: hack/reattach-pv/README.md (hunk 1 above)
Comment (sebgl) [reply hack/reattach-pv/README.md]: You mark a point here. I had in mind the PVs JSON backup could be useful to investigate why something went wrong with the new cluster, by investigating the patched PVs vs. the ones we had before the script ran. But as you point out data is probably already corrupted at this point.

Let's maybe remove the backup feature altogether?

Comment (charith-elastic) [reply hack/reattach-pv/README.md]: Maybe add a line to the README to say that it is the responsibility of the users to backup the underlying disks before attempting to run the tool?

File: hack/reattach-pv/main.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (charith-elastic) [reply hack/reattach-pv/main.go]: Probably not worth the trouble then. An alternative approach I have seen in others tools is that they provide a flag to override the namespace. I am not too bothered about it though. Happy with keeping things as they are.

**Comment by sebgl** [comment]
I removed the JSON backup part and added a warning in the README. @charith-elastic can you get a quick look at the latter?

**Comment by charith-elastic** [comment]
Looks good. Maybe we can be a little bit more forceful and say "You should backup the data in the underlying storage system before attempting to use this tool".  :smile:

**Pull request merged**